### DISCUSSION TITLE
Problems with getting .step() to work

### DISCUSSION THREAD

**Discussion by pascalblanc** [gyscos/cursive#643] [Q&A]
Hi there, 

I admit I am relatively new to all this (programming in general) so sorry if I missed on something very basic.
I tried to run different examples that make use of .step() and, whether it is examples I took from the web or my own, I always encountered
 "method not found in 'Cursive' " at compile time (cursive = 0.16.0 or higher). With cursive = 0.15.0 it still worked. 
 
After some reading of the Cursive documentation, I realised that by **adding this** to my code, I again have access to the siv.step() function and examples compile ok. 

let mut siv = cursive::default()**.into_runner()**;

Is that the correct way of doing this? Maybe there is a "minimal working example" for usage of ".step()" that works with the latest Cursive version?

Thank you very much in advance!
Pascal

**Comment by gyscos** [comment]
Hi!

First, note that using this `step()` function should be considered part of the "advanced" API - in _most_ cases, you don't need it and should be able to just call `.run()`. Especially if you are still learning about Cursive, rust, and programming in general, sticking to the default event loop will make your life much easier.

Now, the details.

Conceptually, a `Cursive` instance just describe the views configuration, the callbacks you configure, and so on. It doesn't _need_ to be running. You _could_ create a `Cursive` object without a terminal, in a daemon, or anything.

Once you actually want to start interacting with a terminal (receiving input and printing stuff) you start what you could call a "session" - this is the `CursiveRunner` object. As long as you have a `CursiveRunner`, it means the terminal is initialized and ready to be used. As soon as you drop it, the terminal is reset back to usual.
While you have this `CursiveRunner`, you can have it just run the event loop by itself with `CursiveRunner::run()`, _or_ you can run it step-by-step yourself. In that case, you would call `CursiveRunner::step()` regularly, [just like its `run` method does](https://docs.rs/cursive_core/0.3.0/src/cursive_core/cursive_run.rs.html#234).

To get this runner, you can indeed use [`CursiveRunnable::into_runner`](https://docs.rs/cursive/latest/cursive/struct.CursiveRunnable.html#method.into_runner). It will initialize the terminal with the default backend, and lets you call this `step()` function yourself.

I agree that this should all be better documented. Right now there's not many examples about it, mainly because it's a rather niche use.

**Reply by pascalblanc** [reply]
Thank you so much for this great and detailed explanation! It really helps me to grasp some concepts of the inner workings of Cursive that apparently I just didn't get before. I am sure I will read your reply a quite few times in the next weeks.😊
Cheers and thanks again.
